qiskit/transpiler/synthesis/aqc/approximate.py

```diff
@@ -0,0 +1,121 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Base classes for an approximate circuit definition."""
+
+from abc import ABC, abstractmethod
+from typing import Optional
+import numpy as np
+
+from qiskit import QuantumCircuit
+
+
+class ApproximateCircuit(QuantumCircuit, ABC):
+    """A base class that represents an approximate circuit."""
+
+    def __init__(self, num_qubits: int, name: Optional[str] = None) -> None:
+        """
+        Args:
+            num_qubits: number of qubit this circuit will span.
+            name: a name of the circuit.
+        """
+        super().__init__(num_qubits, name=name)
+
+    @property
+    def thetas(self) -> np.ndarray:
+        """
+        The property is not implemented and raises a ``NotImplementedException`` exception.
+
+        Returns:
+            a vector of parameters of this circuit.
+        """
+        raise NotImplementedError
+
+    @thetas.setter
+    def thetas(self, thetas: np.ndarray) -> None:
+        """
+        The property is not implemented and raises a ``NotImplementedException`` exception.
+
+        Args:
+            thetas: a vector of parameters to be set in this circuit.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def build(self) -> None:
+        """
+        Constructs this circuit out of the parameters(thetas). Parameter values must be set before
+            constructing the circuit.
+        """
+        raise NotImplementedError
+
+
+class ApproximatingObjective(ABC):
+    """
+    A base class for an optimization problem definition. An implementing class must provide at least
+    an implementation of the ``objective`` method. In such case only gradient free optimizers can
+    be used. Both method, ``objective`` and ``gradient``, preferable to have in an implementation.
+    """
+
+    def __init__(self) -> None:
+        # must be set before optimization
+        self._target_matrix = None
```

Comment:
Im not fully sure why `target_matrix` cannot be set at construction time. `compile_unitary` is called like this
```
        approximating_objective = DefaultCNOTUnitObjective(num_qubits, cnots)

        aqc.compile_unitary(
            target_matrix=unitary,
            approximate_circuit=approximate_circuit,
            approximating_objective=approximating_objective,
        )
 ```
 
 Why wouldnt be possible to do something like this?
 
 ```
         approximating_objective = DefaultCNOTUnitObjective(num_qubits, cnots, unitary)

        aqc.compile_unitary(
            target_matrix=unitary,
            approximate_circuit=approximate_circuit,
            approximating_objective=approximating_objective,
        )
```
 
 In this way, you avoid the dependency "set target_matrix` before calling `objective`/`gradient`.

Reply:
I'm also a little puzzled. I appreciate that it's often cheaper to re-use instances of numerical optimizers across calls with different objective parameters, and so I understand the impulse to defer setting the objective / allowing it to be mutable. Continuing that thought, I'd want to defer setting the target all the way up until I actually ask for optimization to happen. What's gained by saving it as part of the optimizer's state, available before and after optimization completes? Why not supply it _only_ through the `.compile_unitary` call?

Reply:
@1ucian0 It won't work this way. At least in the current design. Now, you pass an objective, a template circuit and a unitary to be optimized in the `aqc.compile_unitary()` call. The problem is that the unitary passed is may be updated in `AQC`, now we normalize it before starting off optimization. So, there should be a way to update the matrix in the objective. OK, another option here is to create an objective, a template inside of `aqc.compile_unitary()`. But then you loose flexibility of passing a custom objective/template.

@ecpeterson Sorry, I don't fully understand your comment. The target matrix is set just before optimization starts. Everything is as stateless as possible and deferred till the latest stages.